versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java

```diff
@@ -311,6 +331,179 @@ public ReferenceDeletedResult delete(NamedRef namedRef, Hash hash)
+    history.current(current);
+
+    for (Reference.PreviousPointer previousPointer : reference.previousPointers()) {
+      history.addPrevious(
+          checkCommit(commitLogic, indexesLogic, previousPointer.pointer(), checked));
+    }
+
+    history.commitLogConsistency(
+        checkCommitLog(commitLogic, indexesLogic, headCommitsToScan, current, checked));
+
+    return history.build();
+  }
+
+  /**
+   * Evaluates the state of the commit log for {@link #getReferenceHistory(String, Integer)}, if
+   * requested. This function returns a result that represents the consistency of the commit log
+   * including <em>direct</em> secondary parent commits, the consistency of the "worst" commit.
+   */
+  private CommitConsistency checkCommitLog(
+      CommitLogic commitLogic,
+      IndexesLogic indexesLogic,
+      Integer headCommitsToScan,
+      ReferenceHistory.ReferenceHistoryElement current,
+      Set<ObjId> checked) {
+    if (headCommitsToScan == null || headCommitsToScan <= 0) {
+      return CommitConsistency.NOT_CHECKED;
+    }
+    if (current.commitConsistency() == CommitConsistency.COMMIT_INCONSISTENT) {
+      // The HEAD commit is inconsistent - do not check the log (it can't be more consistent)
+      return CommitConsistency.COMMIT_INCONSISTENT;
+    }
+
+    // Hard coded limit of 100 commits to check. This function will check at max 1000 commits in the
+    // commit log, or until the configured recorded reference-history interval.
+    int toScan = Math.min(headCommitsToScan, 1000);
+    long nowMicros = persist.config().currentTimeMicros();
+    long inconsistencyWindow =
+        TimeUnit.SECONDS.toMicros(persist.config().referencePreviousHeadTimeSpanSeconds());
+
+    CommitObj headCommit;
+    try {
+      headCommit = commitLogic.fetchCommit(hashToObjId(current.pointer()));
+    } catch (ObjNotFoundException e) {
+      return CommitConsistency.COMMIT_INCONSISTENT;
+    }
+    if (headCommit == null) {
+      return CommitConsistency.COMMIT_CONSISTENT;
+    }
+
+    // Start with the HEAD's direct parent, the HEAD itself is checked elsewhere.
+    ObjId commitId = headCommit.directParent();
+
+    // The "initial" result of the commit log cannot be better than the state of the HEAD, so start
+    // with the HEAD's state.
+    CommitConsistency result = current.commitConsistency();
+
+    while (toScan-- > 0) {
+      if (commitId.equals(EMPTY_OBJ_ID)) {
+        break;
+      }
+
+      CommitObj currentCommit;
+
+      try {
+        currentCommit = commitLogic.fetchCommit(commitId);
+        if (currentCommit == null) {
+          break;
+        }
+      } catch (ObjNotFoundException e) {
+        return CommitConsistency.COMMIT_INCONSISTENT;
+      }
+
+      if (nowMicros - currentCommit.created() > inconsistencyWindow) {
+        break;
+      }
+
+      ReferenceHistory.ReferenceHistoryElement elem =
+          checkCommit(commitLogic, indexesLogic, commitId, checked);
```

Comment:
There is a bit of redundancy here: `checkCommit` will again fetch the commit by id, which is already done above on line 413. I know the cache is supposed to kick in, but I wonder if there could be a way to refactor this and reduce the number of redundant calls to `commitLogic.fetchCommit`.

Reply:
I know - but let's make it "perfect" later. The cache helps here a lot.